plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments.hs

```diff
@@ -0,0 +1,148 @@
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId =
+  (defaultPluginDescriptor plId)
+    { pluginCodeActionProvider = Just codeActionProvider
+    }
+
+haddockCommentsId :: CommandId
+haddockCommentsId = "HaddockCommentsCommand"
+
+codeActionProvider :: CodeActionProvider
+codeActionProvider _lspFuncs ideState _pId (TextDocumentIdentifier uri) range _context =
+  do
+    let nfp = uriToNormalizedFilePath $ toNormalizedUri uri
+    (join -> pm) <- runAction "HaddockComments.GetParsedModule" ideState $ use GetParsedModule `traverse` nfp
+    let locDecls = hsmodDecls . unLoc . pm_parsed_source <$> pm
+        anns = relativiseApiAnns <$> (pm_parsed_source <$> pm) <*> (pm_annotations <$> pm)
+        edits = [runGenComments locDecls anns range gen | gen <- genList]
+    return $ Right $ List [CACodeAction $ toAction "Generate haddock comments" uri edit | (Just edit) <- edits]
+
+genList :: [GenComments]
+genList =
+  [ genForSig,
+    genForRecord
+  ]
+
+-----------------------------------------------------------------------------
+data GenComments = forall a.
+  GenComments
+  { fromDecl :: HsDecl GhcPs -> Maybe a,
+    collectKeys :: a -> [AnnKey],
+    fresh :: Annotation -> Bool,
+    updateAnn :: Annotation -> Annotation
+  }
+
+runGenComments :: Maybe [LHsDecl GhcPs] -> Maybe Anns -> Range -> GenComments -> Maybe TextEdit
+runGenComments mLocDecls mAnns range GenComments {..}
+  | Just locDecls <- mLocDecls,
+    Just anns <- mAnns,
+    [(locDecl, src, x)] <- [(locDecl, l, x) | locDecl@(L l (fromDecl -> Just x)) <- locDecls, inRange range l],
+    annKeys <- collectKeys x,
+    and $ maybe False fresh . flip Map.lookup anns <$> annKeys,
+    anns' <- foldr (Map.adjust updateAnn) anns annKeys,
+    Just range' <- calcRange src range,
+    result <- T.strip . T.pack $ exactPrint locDecl anns' =
+    Just $ TextEdit range' result
+  | otherwise = Nothing
+
+-----------------------------------------------------------------------------
+
+genForSig :: GenComments
+genForSig = GenComments {..}
+  where
+    fromDecl (SigD _ (TypeSig _ _ (HsWC _ (HsIB _ x)))) = Just x
+    fromDecl _ = Nothing
+    updateAnn x = x {annEntryDelta = DP (0, 1), annsDP = dp}
+
+    fresh Ann {annsDP}
+      | null [() | (AnnComment _, _) <- annsDP] = True
+      | otherwise = False
+
+    collectKeys = keyFromTyVar 0
+
+    comment = mkComment "-- ^ ____" noSrcSpan
+    dp = [(AnnComment comment, DP (0, 1)), (G AnnRarrow, DP (1, 2))]
+
+genForRecord :: GenComments
+genForRecord = GenComments {..}
+  where
+    fromDecl (TyClD _ DataDecl {tcdDataDefn = HsDataDefn {dd_cons = cons}}) = Just [x | (L _ ConDeclH98 {con_args = x}) <- cons]
+    fromDecl _ = Nothing
+
+    updateAnn x = x {annEntryDelta = DP (1, -7), annPriorComments = [(comment, DP (1, -7))]}
+
+    fresh Ann {annPriorComments} = null annPriorComments
+
+    collectKeys = keyFromCon
+
+    comment = mkComment "-- | ____" noSrcSpan
```

Comment:
Nice, this feature seems useful. But I personally would prefer if there was no "____" added to each comment, so that I don't have to delete those four characters for every field I want to document. What would you say about generating only the minimal "skeleton" comment required?